connect/api/src/main/java/org/apache/kafka/connect/data/Values.java

```diff
@@ -0,0 +1,1168 @@
+                case FLOAT32:
+                    if (newType == Type.INT8 || newType == Type.INT16 || newType == Type.INT32 || newType == Type.INT64) {
+                        return previous;
+                    }
+                    if (newType == Type.FLOAT64) {
+                        return newSchema;
+                    }
+                    break;
+                case FLOAT64:
+                    if (newType == Type.INT8 || newType == Type.INT16 || newType == Type.INT32 || newType == Type.INT64 || newType ==
+                                                                                                                           Type.FLOAT32) {
+                        return previous;
+                    }
+                    break;
+            }
+            return null;
+        }
+        if (previous.isOptional() == newSchema.isOptional()) {
+            // Use the optional one
+            return previous.isOptional() ? previous : newSchema;
+        }
+        if (!previous.equals(newSchema)) {
+            return null;
+        }
+        return previous;
+    }
+
+    protected static List<Object> alignListEntriesWithSchema(Schema schema, List<Object> input) {
+        if (schema == null) {
+            return input;
+        }
+        Schema valueSchema = schema.valueSchema();
+        if (!isProjectable(valueSchema)) {
+            return input;
+        }
+        List<Object> result = new ArrayList<>();
+        for (Object value : input) {
+            Object newValue = project(valueSchema, value);
+            result.add(newValue);
+        }
+        return result;
+    }
+
+    protected static Map<Object, Object> alignMapKeysAndValuesWithSchema(Schema mapSchema, Map<Object, Object> input) {
+        if (mapSchema == null) {
+            return input;
+        }
+        Schema keySchema = mapSchema.keySchema();
+        Schema valueSchema = mapSchema.valueSchema();
+        if (!isProjectable(keySchema) && !isProjectable(valueSchema)) {
+            return input;
+        }
+        Map<Object, Object> result = new LinkedHashMap<>();
+        for (Map.Entry<?, ?> entry : input.entrySet()) {
+            Object newKey = project(keySchema, entry.getKey());
+            Object newValue = project(valueSchema, entry.getValue());
+            result.put(newKey, newValue);
+        }
+        return result;
+    }
+
+    protected static boolean isProjectable(Schema schema) {
+        switch (schema.type()) {
+            case BYTES:
+                if (Decimal.LOGICAL_NAME.equals(schema.name())) {
+                    return true;
+                }
+                return false;
+            case INT8:
+            case INT16:
+            case INT32:
+            case INT64:
+            case FLOAT32:
+            case FLOAT64:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    protected static Object project(Schema desiredSchema, Object value) {
```

Comment:
This seems a bit repetitive, but more restrictive, as what `convertTo` is doing. Do we actually need both versions? Could we just unify the code with some flag for `project` vs `promote`?

Reply:
No, we don't need the two different forms. That was leftover from an earlier state. I've added another commit that cleans this up.